ops/_tracing/api.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""FIXME docstring."""
+
+from __future__ import annotations
+
+from typing import Any
+
+import ops
+
+from .vendor.charms.certificate_transfer_interface.v1.certificate_transfer import (
+    CertificateTransferRequires,
+)
+from .vendor.charms.tempo_coordinator_k8s.v0.tracing import TracingEndpointRequirer
+
+# NOTE: the shape of certificate_transfer is `{"certificates": list[str]}`
+# it is trivial to read the certificates out manually
+
+
+class Tracing(ops.Object):
+    """FIXME docstring."""
+
+    _certificate_transfer: CertificateTransferRequires | None
+
+    def __init__(
+        self,
+        charm: ops.CharmBase,
+        tracing_relation_name: str,
+        ca_relation_name: str | None = None,
+        ca_data: str | None = None,
+    ):
+        """FIXME docstring.
+
+        Include `ops[tracing]` in your dependencies.
+
+        Declare the relations that the charm supports:
+
+        ```yaml
+        requires:
+            charm-tracing:
+                interface: tracing
+                limit: 1
+            send-ca-cert:
+                interface: certificate_transfer
+                limit: 1
+        ```
+
+        Initialise the tracing thing (FIXME) with the names of these relations.
+
+        ```py
+        import ops.tracing
```

Comment:
Does this actually work at the moment? I can't find any import in `ops/__init__.py` and there doesn't seem to be an `ops/tracing.py` that imports things from `ops/_tracing.py`. Is there some magic I'm missing?

If we're only exposing a `Tracing` and `tracer` name, then I think we would want to (conditionally) have those available in the `ops` namespace, so `ops.tracer` and `ops.Tracing` rather than `ops.tracing.tracer` and `ops.tracing.Tracing`. With Scenario, we already had `ops.testing` as a namespace, and there are a lot of names there, only needed for testing.

Reply:
I've written that with the expectation of the monorepo-style refactor, where:
- a separate package holds the implementation
- `ops.tracing` re-exports the API

Reply:
P.S. `tracer` is or should be private, it's the ops tracer, not the charm tracer.

A "tracer" is an object with attributes like name (ops) and version, thus typically each large library has its own. There are no fixed rules there; my 2c is for coarse tracers, e.g.:
- ops
- charm
- [for example] traefik charm lib